### PR TITLE
DrivAerML: T_max=20 cosine schedule (shorter cycles)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

T_max=20 provides more frequent basin-hopping opportunities than T_max=30. More LR restarts per epoch-budget may help explore the loss landscape more broadly.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 20 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm-tmax-sweep
```

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch
- Test at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT hinata:

## Results

### DrivAerML: T_max=20 cosine schedule (shorter cycles)

| Metric | T_max=20 (this run) | Baseline T_max=30 (PR #2898) |
|--------|---------------------|------------------------------|
| **Best val surface_rel_l2_pct** | **11.06%** @ ep76 | **3.997%** @ ep467 |
| W&B run | t50xj6j8 | bht6h42t |
| Config | 4L/512d/8H, AdamW lr=5e-4 | 4L/512d/8H, AdamW lr=5e-4 |
| Cosine T_max | 20 | 30 |

**External targets:** AB-UPT 3.82% | Transolver-3: 3.71%

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 20 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group dm-tmax-sweep --agent hinata --wandb-name "hinata/dm-tmax-20"
```

### Peak memory
~6.2 GB VRAM on H100 (GPU 0)

### What happened — HYPOTHESIS FALSIFIED

**T_max=20 is catastrophically unstable for DrivAerML without gradient clipping.** The more frequent LR peaks (every 10 epochs vs 15 for T_max=30) trigger periodic gradient explosions that destroy training progress.

**Instability timeline:**
- **ep1-44:** Healthy descent. Best val 11.68% at ep44. Clean trough-to-trough progression: 28.29% → 20.73% → 17.63% → 14.23% → 11.68%.
- **ep46 (spike 1):** Val jumped to 58.75%, grad norm 3.3. Model recovered by ep55 and pushed to new best 11.38%.
- **ep60 (spike 2):** Val jumped to 73.28%, grad norm 2.6. Recovery slower — took ~15 epochs. Pushed to new all-time best **11.06% at ep76**.
- **ep81 (spike 3):** Val jumped to 51.40%, grad norm 2.2. Followed by secondary explosion at ep84 (grad norm 6.1) and ep86 (grad norm 52.6). Recovery took 30+ epochs and never reached pre-spike levels.
- **ep118 (spike 4):** Val jumped to 47.01%, grad norm 11.7. Secondary explosion at ep123 (grad norm 73.0) and ep134 (grad norm 193.6). Model entered permanently degraded state.
- **ep155 (spike 6):** Val 46.71%, grad norm 89.8. Model is in chronic instability — grad norms have not been below 20 since ep118.

**Key finding:** T_max=20 creates more LR peak opportunities per epoch budget, but each peak is a chance for catastrophic gradient explosion. The baseline T_max=30 ran 467 epochs without these issues. The shorter cycle length means less time in the stable low-LR regime between restarts, leaving the model vulnerable to instability at each peak.

**The "more basin-hopping" hypothesis fails because the hopping destroys the model.** Without grad-clip, T_max=20 cannot sustain the long training (400+ epochs) needed to reach competitive DrivAerML metrics. The baseline T_max=30 avoids these explosions and reaches 3.997% — nearly 3x better.

### Suggested follow-ups
1. **T_max=20 + grad-clip=1.0**: The instability is purely a gradient explosion problem. Adding grad-clip might salvage T_max=20's faster cycling while preventing the catastrophic spikes.
2. **T_max=25**: Intermediate value between 20 and 30 — might get some of the basin-hopping benefit without the extreme instability.
3. **T_max=20 + cosine warm restarts with T_mult**: Gradually increasing cycle length could combine frequent early exploration with later stability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3081 because ADVISOR: Closing — T_max=20 is a dead end on DrivAerML without gc. Best val 11.055% at ep76 (177% above 3.997% baseline), then entered permanent degradation with 13 spike events and grad norms up to 193.6. Combined with T_max=15/40/50/100 failures, this conclusively shows T_max=30 is the only stable cosine period for DrivAerML without gradient clipping. Good instability characterization, hinata..

**Comment by tcapelle** [comment]
STUDENT hinata:

## Final Results (run completed — 528 epochs, 360 min)

The run completed the full 360-minute budget. Final state confirms catastrophic divergence:

| Metric | T_max=20 (this run) | Baseline T_max=30 (PR #2898) |
|--------|---------------------|------------------------------|
| **Best val surface_rel_l2_pct** | **11.06%** @ ep76 | **3.997%** @ ep467 |
| **Test at best val checkpoint** | **11.82%** | **6.244%** (PR #2648) |
| Total epochs | 528 | 467 |
| Final val (terminal) | ~58-66% | 3.997% |
| Terminal grad norm | ~300M-5.6B | stable |

The model achieved more total epochs than the baseline (528 vs 467) but the last ~450 epochs were wasted in spike-recovery cycles. By ep134, grad norms had permanently left the healthy range and the model was effectively training on noise for the final 400 epochs.

**Test at best checkpoint: 11.82%** confirms the model never approached competitive performance.
